File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/zip/ZipEntryHeader.java
/*
 * Copyright (c) 2023-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.zip;

import com.ohos.hapsigntool.error.ZipException;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;

/**
 * resolve zip ZipEntryHeader data
 * end of central dir signature    4 bytes  (0x06054b50)
 * number of this disk             2 bytes
 * number of the disk with the
 * start of the central directory  2 bytes
 * total number of entries in the
 * central directory on this disk  2 bytes
 * total number of entries in
 * the central directory           2 bytes
 * size of the central directory   4 bytes
 * offset of start of central
 * directory with respect to
 * the starting disk number        4 bytes
 * .ZIP file comment length        2 bytes
 * .ZIP file comment       (variable size)
 *
 * @since 2023/12/02
 */
public class ZipEntryHeader {
    /**
     * ZipEntryHeader invariable bytes length
     */
    public static final int HEADER_LENGTH = 30;

    /**
     * 4 bytes , entry header signature
     */
    public static final int SIGNATURE = 0x04034b50;

    /**
     * 2 bytes
     */
    private short version;

    /**
     * 2 bytes
     */
    private short flag;

    /**
     * 2 bytes
     */
    private short method;

    /**
     * 2 bytes
     */
    private short lastTime;

    /**
     * 2 bytes
     */
    private short lastDate;

    /**
     * 4 bytes
     */
    private int crc32;

    /**
     * 4 bytes
     */
    private long compressedSize;

    /**
     * 4 bytes
     */
    private long unCompressedSize;

    /**
     * 2 bytes
     */
    private int fileNameLength;

    /**
     * 2 bytes
     */
    private int extraLength;

    /**
     * n bytes
     */
    private String fileName;

    /**
     * n bytes
     */
    private byte[] extraData;

    private int length;

    /**
     * updateLength
     */
    public void updateLength() {
        length = HEADER_LENGTH + fileNameLength + extraLength;
    }

    /**
     * get Zip Entry Header
     *
     * @param bytes ZipEntryHeader bytes
     * @return ZipEntryHeader
     * @throws ZipException read entry header exception
     */
    public static ZipEntryHeader getZipEntryHeader(byte[] bytes) throws ZipException {
        ZipEntryHeader entryHeader = new ZipEntryHeader();
        ByteBuffer bf = ByteBuffer.wrap(bytes);
        bf.order(ByteOrder.LITTLE_ENDIAN);
        if (bf.getInt() != ZipEntryHeader.SIGNATURE) {
            throw new ZipException("find zip entry head failed");
        }
        entryHeader.setVersion(bf.getShort());
        entryHeader.setFlag(bf.getShort());
        entryHeader.setMethod(bf.getShort());
        entryHeader.setLastTime(bf.getShort());
        entryHeader.setLastDate(bf.getShort());
        entryHeader.setCrc32(bf.getInt());
        entryHeader.setCompressedSize(UnsignedDecimalUtil.getUnsignedInt(bf));
        entryHeader.setUnCompressedSize(UnsignedDecimalUtil.getUnsignedInt(bf));
        entryHeader.setFileNameLength(UnsignedDecimalUtil.getUnsignedShort(bf));
        entryHeader.setExtraLength(UnsignedDecimalUtil.getUnsignedShort(bf));
        entryHeader.updateLength();
        return entryHeader;
    }

    /**
     * set entry header name
     *
     * @param bytes name bytes
     */
    public void readFileName(byte[] bytes) {
        ByteBuffer bf = ByteBuffer.wrap(bytes);
        bf.order(ByteOrder.LITTLE_ENDIAN);
        if (fileNameLength > 0) {
            byte[] nameBytes = new byte[fileNameLength];
            bf.get(nameBytes);
            this.fileName = new String(nameBytes, StandardCharsets.UTF_8);
        }
    }

    /**
     * set entry header  extra
     *
     * @param bytes extra bytes
     */
    public void readExtra(byte[] bytes) {
        ByteBuffer bf = ByteBuffer.wrap(bytes);
        bf.order(ByteOrder.LITTLE_ENDIAN);
        if (extraLength > 0) {
            byte[] extra = new byte[extraLength];
            bf.get(extra);
            this.extraData = extra;
        }
    }

    /**
     * change Zip Entry Header to bytes
     *
     * @return bytes
     */
    public byte[] toBytes() {
        ByteBuffer bf = ByteBuffer.allocate(length).order(ByteOrder.LITTLE_ENDIAN);
        bf.putInt(SIGNATURE);
        bf.putShort(version);
        bf.putShort(flag);
        bf.putShort(method);
        bf.putShort(lastTime);
        bf.putShort(lastDate);
        bf.putInt(crc32);
        UnsignedDecimalUtil.setUnsignedInt(bf, compressedSize);
        UnsignedDecimalUtil.setUnsignedInt(bf, unCompressedSize);
        UnsignedDecimalUtil.setUnsignedShort(bf, fileNameLength);
        UnsignedDecimalUtil.setUnsignedShort(bf, extraLength);
        if (fileNameLength > 0) {
            bf.put(fileName.getBytes(StandardCharsets.UTF_8));
        }
        if (extraLength > 0) {
            bf.put(extraData);
        }
        return bf.array();
    }

    public static int getHeaderLength() {
        return HEADER_LENGTH;
    }

    public static int getSIGNATURE() {
        return SIGNATURE;
    }

    public short getVersion() {
        return version;
    }

    public void setVersion(short version) {
        this.version = version;
    }

    public short getFlag() {
        return flag;
    }

    public void setFlag(short flag) {
        this.flag = flag;
    }

    public short getMethod() {
        return method;
    }

    public void setMethod(short method) {
        this.method = method;
    }

    public short getLastTime() {
        return lastTime;
    }

    public void setLastTime(short lastTime) {
        this.lastTime = lastTime;
    }

    public short getLastDate() {
        return lastDate;
    }

    public void setLastDate(short lastDate) {
        this.lastDate = lastDate;
    }

    public int getCrc32() {
        return crc32;
    }

    public void setCrc32(int crc32) {
        this.crc32 = crc32;
    }

    public long getCompressedSize() {
        return compressedSize;
    }

    public void setCompressedSize(long compressedSize) {
        this.compressedSize = compressedSize;
    }

    public long getUnCompressedSize() {
        return unCompressedSize;
    }

    public void setUnCompressedSize(long unCompressedSize) {
        this.unCompressedSize = unCompressedSize;
    }

    public int getFileNameLength() {
        return fileNameLength;
    }

    public void setFileNameLength(int fileNameLength) {
        this.fileNameLength = fileNameLength;
    }

    public int getExtraLength() {
        return extraLength;
    }

    public void setExtraLength(int extraLength) {
        this.extraLength = extraLength;
    }

    public String getFileName() {
        return fileName;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }

    public byte[] getExtraData() {
        return extraData;
    }

    public void setExtraData(byte[] extraData) {
        this.extraData = extraData;
    }

    public int getLength() {
        return length;
    }

    public void setLength(int length) {
        this.length = length;
    }
}

File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/zip/src/zip_entry_header.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "zip_entry_header.h"
#include "signature_tools_log.h"

namespace OHOS {
namespace SignatureTools {
ZipEntryHeader* ZipEntryHeader::GetZipEntryHeader(const std::string& bytes)
{
    ZipEntryHeader* entryHeader = new ZipEntryHeader();
    ByteBuffer bf(bytes.c_str(), bytes.size());

    int32_t entryHeaderInt32Value;
    bf.GetInt32(entryHeaderInt32Value);
    if (entryHeaderInt32Value != ZipEntryHeader::SIGNATURE) {
        delete entryHeader;
        SIGNATURE_TOOLS_LOGE("find zip entry head failed");
        return nullptr;
    }

    int16_t entryHeaderInt16Value;
    bf.GetInt16(entryHeaderInt16Value);
    entryHeader->SetVersion(entryHeaderInt16Value);

    bf.GetInt16(entryHeaderInt16Value);
    entryHeader->SetFlag(entryHeaderInt16Value);

    bf.GetInt16(entryHeaderInt16Value);
    entryHeader->SetMethod(entryHeaderInt16Value);

    bf.GetInt16(entryHeaderInt16Value);
    entryHeader->SetLastTime(entryHeaderInt16Value);

    bf.GetInt16(entryHeaderInt16Value);
    entryHeader->SetLastDate(entryHeaderInt16Value);

    bf.GetInt32(entryHeaderInt32Value);
    entryHeader->SetCrc32(entryHeaderInt32Value);

    uint32_t entryHeaderUInt32Value;
    bf.GetUInt32(entryHeaderUInt32Value);
    entryHeader->SetCompressedSize(entryHeaderUInt32Value);

    bf.GetUInt32(entryHeaderUInt32Value);
    entryHeader->SetUnCompressedSize(entryHeaderUInt32Value);

    uint16_t entryHeaderUInt16Value;
    bf.GetUInt16(entryHeaderUInt16Value);
    entryHeader->SetFileNameLength(entryHeaderUInt16Value);

    bf.GetUInt16(entryHeaderUInt16Value);
    entryHeader->SetExtraLength(entryHeaderUInt16Value);

    entryHeader->SetLength(HEADER_LENGTH + entryHeader->GetFileNameLength() + entryHeader->GetExtraLength());

    return entryHeader;
}

void ZipEntryHeader::ReadFileName(const std::string& bytes)
{
    ByteBuffer bf(bytes.c_str(), bytes.size());
    if (m_fileNameLength > 0) {
        std::string nameBytes(m_fileNameLength, 0);
        bf.GetData(&nameBytes[0], m_fileNameLength);
        m_fileName = nameBytes;
    }
}

void ZipEntryHeader::ReadExtra(const std::string& bytes)
{
    ByteBuffer bf(bytes.c_str(), bytes.size());
    if (m_extraLength > 0) {
        std::string extra(m_extraLength, 0);
        bf.GetData(&extra[0], m_extraLength);
        m_extraData = extra;
    }
}

std::string ZipEntryHeader::ToBytes()
{
    ByteBuffer bf(m_length);

    bf.PutInt32(SIGNATURE);
    bf.PutInt16(m_version);
    bf.PutInt16(m_flag);
    bf.PutInt16(m_method);
    bf.PutInt16(m_lastTime);
    bf.PutInt16(m_lastDate);
    bf.PutInt32(m_crc32);
    bf.PutUInt32(m_compressedSize);
    bf.PutUInt32(m_unCompressedSize);
    bf.PutUInt16(m_fileNameLength);
    bf.PutUInt16(m_extraLength);

    if (m_fileNameLength > 0) {
        bf.PutData(m_fileName.c_str(), m_fileName.size());
    }
    if (m_extraLength > 0) {
        bf.PutData(m_extraData.c_str(), m_extraData.size());
    }

    return bf.ToString();
}

int ZipEntryHeader::GetHeaderLength()
{
    return HEADER_LENGTH;
}

int ZipEntryHeader::GetSIGNATURE()
{
    return SIGNATURE;
}

short ZipEntryHeader::GetVersion()
{
    return m_version;
}

void ZipEntryHeader::SetVersion(short version)
{
    m_version = version;
}

short ZipEntryHeader::GetFlag()
{
    return m_flag;
}

void ZipEntryHeader::SetFlag(short flag)
{
    m_flag = flag;
}

short ZipEntryHeader::GetMethod()
{
    return m_method;
}

void ZipEntryHeader::SetMethod(short method)
{
    m_method = method;
}

short ZipEntryHeader::GetLastTime()
{
    return m_lastTime;
}

void ZipEntryHeader::SetLastTime(short lastTime)
{
    m_lastTime = lastTime;
}

short ZipEntryHeader::GetLastDate()
{
    return m_lastDate;
}

void ZipEntryHeader::SetLastDate(short lastDate)
{
    m_lastDate = lastDate;
}

int ZipEntryHeader::GetCrc32()
{
    return m_crc32;
}

void ZipEntryHeader::SetCrc32(int crc32)
{
    m_crc32 = crc32;
}

uint32_t ZipEntryHeader::GetCompressedSize()
{
    return m_compressedSize;
}

void ZipEntryHeader::SetCompressedSize(uint32_t compressedSize)
{
    m_compressedSize = compressedSize;
}

uint32_t ZipEntryHeader::GetUnCompressedSize()
{
    return m_unCompressedSize;
}

void ZipEntryHeader::SetUnCompressedSize(uint32_t unCompressedSize)
{
    m_unCompressedSize = unCompressedSize;
}

uint16_t ZipEntryHeader::GetFileNameLength()
{
    return m_fileNameLength;
}

void ZipEntryHeader::SetFileNameLength(uint16_t fileNameLength)
{
    m_fileNameLength = fileNameLength;
}

uint16_t ZipEntryHeader::GetExtraLength()
{
    return m_extraLength;
}

void ZipEntryHeader::SetExtraLength(uint16_t extraLength)
{
    m_extraLength = extraLength;
}

std::string ZipEntryHeader::GetFileName() const
{
    return m_fileName;
}

void ZipEntryHeader::SetFileName(const std::string& fileName)
{
    m_fileName = fileName;
}

std::string ZipEntryHeader::GetExtraData() const
{
    return m_extraData;
}

void ZipEntryHeader::SetExtraData(const std::string& extraData)
{
    m_extraData = extraData;
}

uint32_t ZipEntryHeader::GetLength()
{
    return m_length;
}

void ZipEntryHeader::SetLength(uint32_t length)
{
    m_length = length;
}
} // namespace SignatureTools
} // namespace OHOS